src/probnum/kernels/_kernel.py

```diff
@@ -1,135 +1,359 @@
+        positive (semi-)definite for ``x0 == x1``.
+
+        Parameters
+        ----------
+        x0 : array-like
+            First set of inputs to the (cross-)covariance function as an array of shape
+            ``(M, D)``, where ``D`` is either 1 or :attr:`input_dim`.
+        x1 : array-like
+            Optional second set of inputs to the (cross-)covariance function as an array
+            of shape ``(N, D)``, where ``D`` is either 1 or :attr:`input_dim`.
+            If ``x1`` is not specified, the function behaves as if ``x1 = x0``.
+
+        Returns
+        -------
+        kernmat : numpy.ndarray
+            The matrix / stack of matrices containing the pairwise evaluations of the
+            (cross-)covariance function(s) on ``x0`` and ``x1`` as an array of shape
+            ``(M, N)`` if :attr:`shape` is ``()`` or
+            ``(S[l - 1], ..., S[1], S[0], M, N)``, where ``S`` is :attr:`shape` if
+            :attr:`shape` is non-empty.
+
+        Raises
+        ------
+        ValueError
+            If the shapes of the inputs don't match the specification.
+
+        See Also
+        --------
+        __call__: Evaluate the kernel more flexibly.
+
+        Examples
+        --------
+        See documentation of class :class:`Kernel`.
         """
+
+        x0 = np.array(x0)
+        x1 = x0 if x1 is None else np.array(x1)
+
+        # Shape checking
+        errmsg = (
+            "`{argname}` must have shape `(N, D)` or `(D,)`, where `D` is the input "
+            f"dimension of the kernel (D = {self.input_dim}), but an array with shape "
+            "`{shape}` was given."
+        )
+
+        if not (1 <= x0.ndim <= 2 and x0.shape[-1] in (self.input_dim, 1)):
+            raise ValueError(errmsg.format(argname="x0", shape=x0.shape))
+
+        if not (1 <= x1.ndim <= 2 and x1.shape[-1] in (self.input_dim, 1)):
+            raise ValueError(errmsg.format(argname="x1", shape=x1.shape))
+
+        # Pairwise kernel evaluation
+        return self(x0[:, None, :], x1[None, :, :])
+
+    @property
+    def input_dim(self) -> int:
+        """Dimension of arguments of the covariance function."""
         return self._input_dim
 
     @property
-    def output_dim(self) -> int:
-        """Dimension of the evaluated covariance function.
+    def shape(self) -> ShapeType:
+        """If :attr:`shape` is ``()``, the :class:`Kernel` instance represents a
+        single (cross-)covariance function. Otherwise, i.e. if :attr:`shape` is
+        non-empty, the :class:`Kernel` instance represents a tensor of
+        (cross-)covariance functions whose shape is given by ``shape``."""
+        return self._shape
 
-        The resulting evaluated kernel :math:`k(x_0, x_1) \\in
-        \\mathbb{R}^{d_{out} \\times d_{out}}` has *shape=(output_dim,
-        output_dim)*.
+    @abc.abstractmethod
+    def _evaluate(
+        self,
+        x0: ArrayLike,
+        x1: Optional[ArrayLike],
```

Comment:
```suggestion
        x1: Optional[ArrayLike]=None,
```

Reply:
See above. Besides, this is never called with default arguments anyway